warehouse/query-core/src/main/java/datawave/query/jexl/visitors/IngestTypePruningVisitor.java

```diff
@@ -165,12 +156,33 @@ public Object visit(ASTReference node, Object data) {
 
     @Override
     public Object visit(ASTReferenceExpression node, Object data) {
-        return visitOrPrune(node, data);
+        Object o = node.jjtGetChild(0).jjtAccept(this, data);
+        if (node.jjtGetNumChildren() == 0) {
+            pruneNodeFromParent(node);
+        }
+        return o;
     }
 
     @Override
     public Object visit(ASTOrNode node, Object data) {
-        return visitOrPrune(node, data);
+
+        if (data == null) {
+            // just a visit
+            return ingestTypeVisitor.getIngestTypesForJunction(node);
+        }
+
+        Set<String> pruningTypes = (Set<String>) data;
+
+        // must traverse the children in reverse order because of pruning
+        for (int i = node.jjtGetNumChildren() - 1; i >= 0; i--) {
+            node.jjtGetChild(i).jjtAccept(this, pruningTypes);
+        }
+
+        if (node.jjtGetNumChildren() == 0) {
```

Comment:
Should this check be placed before the for loop?

Reply:
No, a child of a union can self-prune in certain cases. Consider the query `(A && B) || (C && D)` where each term maps to a separate, unique datatype (no overlaps). The `(A && B)` branch will self prune to zero, and the `(C && D)` branch will prune to zero. Thus the union node now has zero children and must prune itself from it's parent.

Reply:
Can you add a comment to explain that this will only happen in some instances? Do we have a unit test that shows this?

Reply:
I'll add a comment. You can look at the `testDoubleNestedPruning()` unit test for some examples.